libass/ass_render.c

```diff
@@ -684,6 +698,127 @@ static inline size_t bitmap_size(const Bitmap *bm)
     return bm->stride * bm->h;
 }
 
+static void apply_hscroll_fade(RenderContext *state, ASS_Image *head)
+{
+    double fade_width = state->scroll_fade_width * state->screen_scale_x;
+    if (fade_width <= 0)
+        return;
+
+    ASS_Renderer *render_priv = state->renderer;
+
+    for (ASS_Image *cur = head; cur; cur = cur->next) {
+        int event_start = cur->dst_x - render_priv->settings.left_margin;
+        int event_end = event_start + cur->w;
+
+        double leading_fade_start = -0.5;
+        int leading_fade_start_int = floor(leading_fade_start) + 1;
+        double leading_fade_end = leading_fade_start + fade_width;
+        int leading_fade_end_int = ceil(leading_fade_end);
+
+        double trailing_fade_end = render_priv->frame_content_width - 0.5;
+        int trailing_fade_end_int = ceil(trailing_fade_end);
+        double trailing_fade_start = trailing_fade_end - fade_width;
+        int trailing_fade_start_int = floor(trailing_fade_start) + 1;
+
+        int full_a = _a(cur->color);
+        cur->color |= 0xFF;  // starts out fully transparent
+
+        for (int x = FFMAX(event_start, leading_fade_start_int); x < event_end && x <= trailing_fade_end_int; ) {
+            double exact_a = 0xFF - full_a;
+            if (x < leading_fade_end_int)
+                exact_a *= (x - leading_fade_start) / fade_width;
+            if (x >= trailing_fade_start_int)
+                exact_a *= FFMAX(trailing_fade_end - x, 0) / fade_width;
+
+            int a = 0xFF - lround(exact_a);
+            if (a != _a(cur->color)) {
+                if (x > event_start) {
+                    ASS_Image *nimg = ass_image_borrow_copy(cur);
+                    nimg->dst_x = x + render_priv->settings.left_margin;
+                    cur->w = nimg->dst_x - cur->dst_x;
+                    nimg->w -= cur->w;
+                    nimg->bitmap += cur->w;
+                    cur->next = nimg;
+                    cur = nimg;
```

Comment:
That would erase the distinction between shadows, borders and main images for events affected by gradients, defeating the point of `image->type`. We could, of course, make a whole parallel set of _three_ new values… But it would still be impossible in some cases to tell where one gradient event ends and another gradient event starts.

By the way, why do the `IMAGE_TYPE_...` constants not start with `ASS_`? Sigh.

Reply:
> By the way, why do the IMAGE_TYPE_... constants not start with ASS_? Sigh.

Btw, all macros and enums in `ass_types.h` but the `JUSTIFY` ones also lack an `ASS_` prefix.